### PR TITLE
Track the current state of the room

### PR DESCRIPTION
**Pull request by NegativeMjark**
Include the changes to current state caused by an event when writing the event to the room server output log.

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/roomserver/storage/events_table.go [hunk 1]
```diff
@@ -297,6 +295,28 @@ func (s *eventStatements) bulkSelectEventReference(eventNIDs []types.EventNID) (
 	return results, nil
 }
 
+func (s *eventStatements) bulkSelectEventID(eventNIDs []types.EventNID) (map[types.EventNID]string, error) {
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/events_table.go]: Comment the value of the output map please. I assume this is `eventNID -> event_id`.

File: src/github.com/matrix-org/dendrite/roomserver/storage/rooms_table.go [hunk 2]
```diff
@@ -18,7 +18,10 @@ CREATE TABLE IF NOT EXISTS rooms (
     -- (The server will be in that state while it stores the events for the initial state of the room)
     latest_event_nids BIGINT[] NOT NULL DEFAULT '{}'::BIGINT[],
     -- The last event written to the output log for this room.
-    last_event_sent_nid BIGINT NOT NULL DEFAULT 0
+    last_event_sent_nid BIGINT NOT NULL DEFAULT 0,
+	-- The state of the room after the current set of latest events.
+	-- This will be 0 if there are no latest events in the room.
+	state_snapshot_nid BIGINT NOT NULL DEFAULT 0
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/storage/rooms_table.go]: Tabs!

File: src/github.com/matrix-org/dendrite/roomserver/types/types.go [hunk 3]
```diff
@@ -133,6 +133,12 @@ type StateEntryList struct {
 // (On postgresql this wraps a database transaction that holds a "FOR UPDATE"
 //  lock on the row holding the latest events for the room.)
 type RoomRecentEventsUpdater interface {
+	// The latest event IDs and state in the room.
+	LatestEvents() []StateAtEventAndReference
+	// The event ID of the latest event sent in the room.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/types/types.go]: When you say "sent", do you mean:
 - Sent in the room?
 - Send to the output log?

We should be explicit.

File: src/github.com/matrix-org/dendrite/roomserver/input/state.go [hunk 4]
```diff
@@ -10,7 +10,7 @@ import (
 // calculateAndStoreState calculates a snapshot of the state of a room before an event.
 // Stores the snapshot of the state in the database.
 // Returns a numeric ID for that snapshot.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/input/state.go]: - `s/calculateAndStoreState/calculateAndStoreStateBeforeEvent/`
- When you say "returns a numeric ID for that snapshot" that's unhelpful, I can see it's a snapshot NID. More importantly, this snapshot is the state of the room **after** the given event? Before?

File: src/github.com/matrix-org/dendrite/roomserver/input/state.go [hunk 5]
```diff
@@ -107,6 +113,98 @@ func calculateAndStoreStateMany(db RoomEventDatabase, roomNID types.RoomNID, pre
...
+	stateBlockNIDList := stateBlockNIDLists[0]
+
+	stateEntryLists, err := db.StateEntries(stateBlockNIDList.StateBlockNIDs)
+	if err != nil {
+		return nil, err
+	}
+	stateEntriesMap := stateEntryListMap(stateEntryLists)
+
+	// Combined all the state entries for this snapshot.
+	// The order of state data NIDs in the list tells us the order to combine them in.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/input/state.go]: >state data NIDs

File: src/github.com/matrix-org/dendrite/roomserver/storage/events_table.go (hunk 1 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/storage/events_table.go]: Done 875ce20

File: src/github.com/matrix-org/dendrite/roomserver/storage/rooms_table.go (hunk 2 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/storage/rooms_table.go]: Fixed 6acc5f4

File: src/github.com/matrix-org/dendrite/roomserver/types/types.go (hunk 3 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/types/types.go]: Done 57a6ba7

File: src/github.com/matrix-org/dendrite/roomserver/input/state.go (hunk 4 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/input/state.go]: Done e4189ef

File: src/github.com/matrix-org/dendrite/roomserver/input/state.go (hunk 5 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/input/state.go]: Fixed 2763649

File: src/github.com/matrix-org/dendrite/roomserver/input/state.go [hunk 6]
```diff
@@ -25,6 +25,12 @@ func calculateAndStoreState(
 		return 0, err
 	}
 
+	return calculateAndStoreStateAfterEvents(db, roomNID, prevStates)
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/input/state.go]: I don't understand why the return value for `calculateAndStoreStateBeforeEvent` is the same as `calculateAndStoreStateAfterEvents`. Please explain.

Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/input/state.go]: Does ece677c help? I'm not sure about the wording.

**Comment by Kegsay** [comment]
LGTM

**Pull request merged**